Add support for list insertion for mutation removal

penguinwu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54271 Add support for list insertion for mutation removal**

Summary: 

Test Plan: python test/test_jit.py TestRemoveMutation.test_lists_insert

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27180031](https://our.internmc.facebook.com/intern/diff/D27180031)

Commit:
Add support for list insertion for mutation removal

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison commented on test/jit/test_remove_mutation.py:
```diff
@@ -193,6 +194,23 @@ def intermediary_use():
 
+    def test_lists_insert(self):
+        def successful_remove():
+            a = torch.jit.annotate(List[int], [])
```
nit: this is a deprecated api, use `a: List[int] = []`

eellison commented on torch/csrc/jit/passes/remove_mutation.cpp:
```diff
@@ -139,15 +142,37 @@ void MutationRemover::RemoveListMutation(Block* block) {
+    }
+
+    // process use-chain and aliasing of node output
+    bool has_output = (node->outputs().size() > 0);
```
just fyi, neither append nor insert return a value in python so it's basically impossible for the outputs to have uses if people are scripting normal python code. i submitted a PR to fix this but unfortunately onnx relies on it and it doesnt really hurt anything, just kind of confusing, so I didn't push to land the PR: https://github.com/pytorch/pytorch/pull/39536.

either way this is fine here for the potentially confused, even though i think it will never be called

eellison reviewed (approved):
Nice! Looks good

Commit:
Update on "Add support for list insertion for mutation removal"


Summary: 

Test Plan: python test/test_jit.py TestRemoveMutation.test_lists_insert

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "Add support for list insertion for mutation removal"


Summary: 

Test Plan: python test/test_jit.py TestRemoveMutation.test_lists_insert

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27180031](https://our.internmc.facebook.com/intern/diff/D27180031)

[ghstack-poisoned]

Commit:
Update on "Add support for list insertion for mutation removal"


Summary: 

Test Plan: python test/test_jit.py TestRemoveMutation.test_lists_insert

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27180031](https://our.internmc.facebook.com/intern/diff/D27180031)

[ghstack-poisoned]

Commit:
Update on "Add support for list insertion for mutation removal"


Summary: 

Test Plan: python test/test_jit.py TestRemoveMutation.test_lists_insert

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27180031](https://our.internmc.facebook.com/intern/diff/D27180031)

[ghstack-poisoned]
